plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java

```diff
@@ -1456,22 +1496,65 @@ public Answer createSnapshot(final CreateObjectCommand cmd) {
             final KVMPhysicalDisk disk = storagePoolMgr.getPhysicalDisk(primaryStore.getPoolType(), primaryStore.getUuid(), volume.getPath());
             if (state == DomainInfo.DomainState.VIR_DOMAIN_RUNNING && !primaryPool.isExternalSnapshot()) {
                 final String vmUuid = vm.getUUIDString();
-                final Object[] args = new Object[] {snapshotName, vmUuid};
-                final String snapshot = SnapshotXML.format(args);
+                boolean isKVMEnabled = cmd.getContextParam("kvmsnapshot") != null ? Boolean.parseBoolean(cmd.getContextParam("kvmsnapshot")) : false;
+                s_logger.debug(String.format("Snapshots on KVM is enabled %s", isKVMEnabled));
+                if (isKVMEnabled) {
+                    long size = 0;
+                    OutputInterpreter.OneLineParser parser = new OutputInterpreter.OneLineParser();
+                    Script sc = new Script("virsh");
+                    sc.add(String.format("qemu-monitor-command %s '{ \"execute\" : \"query-block\" }'", vmName));
```

Comment:
Same here as my other comments

Reply:
@wido libvirt-java is not updated for quite long time.
'virsh' provides more functions.

Reply:
Yes, but then we fix that. It's a matter of sending a patch to the libvirt project and they are happy to accept it.

I think cloudstack is the only project really using those bindings. But they are not so difficult, so we can send patches when we need something.

We should not just work around it by not using them. Executing commands is just not reliable enough.

Reply:
Thanks @wido for the comments!
We will check what will be needed for the patch, and we will start working on it. We will ask for help, if we get stuck somewhere.
Could you please give me and example why "executing commands is just not reliable enough"? And if there are a problems with commands execution whether there will be the same with execution of virsh in helper scripts?

Reply:
All the calls to libvirt from the agent are done through the Java bindings. If we suddenly start to execute virsh from Java we work around all the existing code which is there to have an open connection with libvirt.

In addition you can only check on exit status and not handle exceptions properly coming back from libvirt.

If we are lacking features in libvirt-java those need to be added through the libvirt project.

Reply:
replaced with libvirt's Java bindings